### DISCUSSION TITLE
Waveshare 1.44 inch LCD display HAT

### DISCUSSION THREAD

**Discussion by vitotai** [outdoorbits/little-backup-box#303] [Show and tell]
I hate soldering, so that the ready-made HAT is a good choice for me. 
 
https://www.waveshare.com/1.44inch-lcd-hat.htm

The driver IC is supported by luma.lcd, so it can be done by adding some code. However, it still took me a while to make it work, given the fact that I don't program in python.

The modification to display.py
1. add import, around line 63
``` 
from luma.lcd.device import st7735
```
2. add some parameters for SPI bus, around line 136. (This seems to be the key.)
```
serial = spi(port=self.conf_DISP_SPI_PORT, device=0, bus_speed_hz=40000000, gpio_DC=25, gpio_RST=27)
````
3. create the display driver. around line 158. (pin=24 is necessary, width and height maybe not.) 
```
			elif self.conf_DISP_DRIVER == "ST7735":
				self.device = st7735(serial,pin=24, width=128, height=128)
```

Surely you will need to modify "setup.php" to make st7735 available, as well as 128 vertical resolution. It's better to have DC, RST, and BL pin as options, but hard-coding is easier for me.

In setup page, choose SPI and port 0. 

According to the document, the pins of joystick matches button menu configuration #3. However, the configuration seems to have joystick on the left-hand side, which might need to rotate the display. Therefore, I add custom configurations in /var/www/little-backup-box/media/buttons.private.cfg
```
19=up,6=down,26=left,5=right:Joystick
19=up,6=down,21=left,16=right:button
```

So, by choosing "c1", the joystick is left-hand sided.  I prefer using the other buttons for "enter" and "exit" instead of "right" and "left". The "c2" configuration will do so.

Just ask if more information is needed.

**Comment by outdoorbits** [comment]
Great, I'll integrate this as soon as possible.

**Comment by outdoorbits** [comment]
Am I right, backlight is pin 24 = GPIO8 (CE0)?

**Reply by outdoorbits** [reply]
I just published an update. You now can select the waveshare display and the button variants. Would you please test it? Thank you!

**Comment by vitotai** [comment]
tested it.

The "ST7735 WAVESHARE LCD display HAT" didn't work. Maybe the spaces have something to do with?
ST7735 worked fine. However, 128 horizontal resolution is missed.

BTW,
other settings for Waveshare HAT
- Color: RGB
- Button edge Detect: Falling
- Pull resistor: UP

**Comment by vitotai** [comment]
I forget the "BGR" parameter:
```
self.device = st7735(serial,pin=24, width=128, height=128, bgr=True)
```

**Comment by outdoorbits** [comment]
I fixed this. I removed "width=128, height=128", it is set now by 
self.device.capabilities(width=self.conf_DISP_RESOLUTION_X, height=self.conf_DISP_RESOLUTION_Y, rotate=self.conf_DISP_ROTATE, mode=self.conf_DISP_COLOR_MODEL)
like any other display. Please set the resolution in hardware. Excited about your test...

**Reply by vitotai** [reply]
This update didn't work for me. Neither st7735, nor "ST7735 WAVESHARE LCD display HAT".

**Reply by outdoorbits** [reply]
Did you configure the resolution in hardware settings?

**Comment by outdoorbits** [comment]
Can you figure out the problem?

**Reply by vitotai** [reply]
I tried again, and it seems to work fine. 
Maybe I've made some mistake. 
BTW, there was a small band at the bottom of the display. After googling, offset parameters are needed.
```
self.device = st7735(serial,pin=24, h_offset=1, v_offset=2, bgr=True)
```

**Comment by outdoorbits** [comment]
Offset can be configured in settings now. My display seems not to support this. Does it work for you?

**Reply by vitotai** [reply]
The offset settings work for me.

FYI.Here is the source that I found the information.
https://github.com/rm-hull/luma.lcd/issues/70

**Comment by outdoorbits** [comment]
Please test again, I merged a huge part of code and hope, nothing is 
broken now... If it's still o.k., I would close this. Thank you!

**Reply by vitotai** [reply]
It works flawlessly.

**Reply by outdoorbits** [reply]
Thank you!
